typing/typecore.ml

```diff
@@ -778,21 +791,66 @@ let solve_constructor_annotation
         | _ -> assert false
   in
   if ids <> [] then ignore begin
+    let get_decl_manifest id env =
+      let decl =
+        try Env.find_type (Path.Pident id) env with Not_found -> assert false in
+      match decl.type_manifest with
+        None -> assert false
+      | Some ty -> (decl, ty)
+    in
     let ids = List.map (fun x -> x.txt) ids in
     let rem =
+      (* First process the existentials introduced by this constructor.
+         Just need to make their definitions abstract. *)
       List.fold_left
         (fun rem tv ->
           match get_desc tv with
             Tconstr(Path.Pident id, [], _) when List.mem id rem ->
+              let decl, tv' = get_decl_manifest id !!penv in
+              let env =
+                Env.add_type ~check:false id
+                  {decl with type_manifest = None} !!penv
+              in
+              Pattern_env.set_env penv env;
+              (* We have changed the definition, so clean up *)
+              Btype.cleanup_abbrev ();
+              (* Since id is now abstract, this does not create a cycle *)
+              unify_pat_types cty.ctyp_loc env tv tv';
               list_remove id rem
           | _ ->
               raise (Error (cty.ctyp_loc, !!penv,
                             Unbound_existential (ids, ty))))
         ids ty_ex
     in
-    if rem <> [] then
-      raise (Error (cty.ctyp_loc, !!penv,
-                    Unbound_existential (ids, ty)))
+    (* The other type names should be bound to newly introduced existentials. *)
+    let bound_ids = ref ids in
+    List.iter
+      (fun id ->
+        let decl, tv' = get_decl_manifest id !!penv in
+        let tv' = expand_head !!penv tv' in
+        begin match get_desc tv' with
+        | Tconstr (Path.Pident id', [], _) when
+            Ident.scope id' = penv.equations_scope ->
+              if List.exists (Ident.same id') !bound_ids then
+                raise (Error (cty.ctyp_loc, !!penv,
+                              Bind_existential (Bind_already_bound, id, tv')));
+              (* Both id and id' are Scoped identifiers, so their stamps grow *)
+              if Ident.compare_stamp id id' > 0 then
+                raise (Error (cty.ctyp_loc, !!penv,
+                              Bind_existential (Bind_not_in_scope, id, tv')));
+              bound_ids := id' :: !bound_ids
+        | _ ->
+            raise (Error (cty.ctyp_loc, !!penv,
+                          Bind_existential
+                            (Bind_non_locally_abstract, id, tv')));
+        end;
+        let env =
+          Env.add_type ~check:false id
+            {decl with type_manifest = Some (correct_levels tv')} !!penv
+        in
```

Comment:
The logic between the two loops (on existential instances, then on remaining user-defined names) is fairly different: in one case you create a new abstract (for the user-provided type name) and you unify, and in the other case you define the user-provided name as a synonym to the typer-generated existential. Can you explain why there is this difference? Will it make a difference for error messages? (In particular, for not-purely-existential names, are they shown in error messages in the right-hand-side of the clause with the name given by the user?)

Reply:
As I explained above, in the latter case we do not replace the original name by the one given by the user, but just define an alias, so we cannot guarantee that the internal name will never show up. In short-path mode we could tune the heuristics to show the user-defined name.